modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/CogGranuleAccessProvider.java

```diff
@@ -0,0 +1,127 @@
+import javax.imageio.spi.ImageReaderSpi;
+import javax.imageio.stream.ImageInputStream;
+import org.geotools.coverage.grid.io.AbstractGridCoverage2DReader;
+import org.geotools.coverage.grid.io.imageio.MaskOverviewProvider;
+import org.geotools.coverage.grid.io.imageio.MaskOverviewProvider.SpiHelper;
+import org.geotools.gce.imagemosaic.catalog.CatalogConfigurationBean;
+import org.geotools.gce.imagemosaic.catalog.CogConfigurationBean;
+import org.geotools.util.Utilities;
+import org.geotools.util.factory.Hints;
+
+import static org.geotools.gce.imagemosaic.GranuleDescriptor.EXCLUDE_MOSAIC;
+
+/** GranuleAccessProvider for COG Granules. */
+class CogGranuleAccessProvider extends DefaultGranuleAccessProvider
+        implements GranuleAccessProvider {
+
+    private static final ImageReaderSpi DEFAULT_COG_IMAGE_READER_SPI = new CogImageReaderSpi();
+    private static final ImageInputStreamSpi DEFAULT_COG_IMAGE_INPUT_STREAM_SPI =
+            new CogImageInputStreamSpi();
+
+    private static final Logger LOGGER =
+            org.geotools.util.logging.Logging.getLogger(CogGranuleAccessProvider.class);
+
+    /** A catalog bean containing details o */
+    private CogConfigurationBean cogConfig;
+
+    public CogGranuleAccessProvider(CatalogConfigurationBean bean) {
+        // A Cog Provider will always have at least a streamSpi and an ImageReaderSpi
+        this(getHints(bean));
+        cogConfig = bean.getCogBean();
+    }
+
+    private static Hints getHints(CatalogConfigurationBean bean) {
+        // A Cog provider will always have an ImageReaderSpi and an ImageInputStream
+        // Using defaults COGs when not specified
+        Utilities.ensureNonNull("CatalogConfigurationBean", bean);
+        Hints hints = new Hints();
+        ImageReaderSpi readerSpi = createImageReaderSpiInstance(bean.getSuggestedSPI());
+        if (readerSpi == null) {
+            readerSpi = DEFAULT_COG_IMAGE_READER_SPI;
+        }
+        hints.put(GranuleAccessProvider.SUGGESTED_READER_SPI, readerSpi);
+        ImageInputStreamSpi imageInputStreamSpi =
+                createImageInputStreamSpiInstance(bean.getSuggestedIsSPI());
+        if (imageInputStreamSpi == null) {
+            imageInputStreamSpi = DEFAULT_COG_IMAGE_INPUT_STREAM_SPI;
+        }
+        hints.put(GranuleAccessProvider.SUGGESTED_STREAM_SPI, imageInputStreamSpi);
+        hints.add(EXCLUDE_MOSAIC);
+        return hints;
+    }
+
+    public CogGranuleAccessProvider(Hints hints) {
+        super(hints);
+    }
+
+    @Override
+    public void setGranuleInput(Object input) throws IOException {
+        this.inputUrl = (URL) input;
+        CogUri cogUri = cogConfig.createUri(inputUrl.toString());
+        this.input =
+                new CogSourceSPIProvider(
+                        cogUri, imageReaderSpi, imageInputStreamSpi, cogConfig.getRangeReader());
+    }
+
+    @Override
+    public MaskOverviewProvider getMaskOverviewsProvider() throws IOException {
```

Comment:
This seems to duplicate the entire parent method for a boolean flag in the MaskOverviewProvider. Maybe have a protected method that takes the boolean in the base class, and call it from the base class and from this class with different boolean values instead?

Reply:
Actually the real difference is in the SpiHelper constructor using a SourceSPIProvider in the COG implementation whilst the default one uses a URL.

Reply:
One thing that came to mind after the review. Checking the external files is expensive in other occasions, could it be a configuration parameter? And maybe if cog is enabled, you just raise it by default unless explicitly specified? (convenient, but magic, should be well documented). This would make the COG path less special.

Reply:
I can add it as a config param so it will "override" the System.property right?